sds: Ignore Delta Removals

keithmattix opened the pull request:
<!--
!!!ATTENTION!!!

If you are fixing *any* crash or *any* potential security issue, *do not*
open a pull request in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.
Thank you in advance for helping to keep Envoy secure.

!!!ATTENTION!!!

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)
-->

Commit Message: Ignore Delta SDS removals
Additional Description: Fixes #24373 and #32832
Risk Level: Low
Testing: Manual testing with the Istio scenario described in #32832. Investigating how to add a unit test
Docs Changes:
Release Notes: Delta SDS removals will no longer result in a "Missing SDS resources" error message
Platform Specific Features:

adisuissa commented on source/common/secret/sds_api.cc:
```diff
@@ -154,14 +168,17 @@ void SdsApi::onConfigUpdateFailed(Envoy::Config::ConfigUpdateFailureReason reaso
-  if (num_resources == 0) {
+absl::Status SdsApi::validateUpdateSize(uint32_t added_resources_num,
+                                        uint32_t removed_resources_num) {
+  if (added_resources_num + removed_resources_num == 0) {
```
nit: prefer `(A == 0 && B == 0)`, to avoid overflow (not that it should ever happen in reality, but just to be on the safe side).

adisuissa commented on source/common/secret/sds_api.cc:
```diff
@@ -154,14 +168,17 @@ void SdsApi::onConfigUpdateFailed(Envoy::Config::ConfigUpdateFailureReason reaso
+                    "{}, number of removed resources {}",
+                    sds_config_name_, added_resources_num, removed_resources_num));
   }
   return absl::OkStatus();
```
hmm... this seems to allow adding and removing of the resource (which I guess should just be updating the resource) when both added and removed are 1.
To the best of my (limited) understanding SDS should currently be a single-resource subscription. That implies that for a given resource name it should either be in the added part or removed part.

If you think this is incorrect, please point me to a test or config that you are aware of.

keithmattix replied:
That was my understanding as well based on the code, but I came across this [test](https://github.com/envoyproxy/envoy/blob/ffcc257e16c9046b2fec7497a6bf9293d8ada286/test/common/secret/sds_api_test.cc#L562) that seemed to imply add + removal is allowed? I'm open to modifying the test though (with a release note clarifying the behavior)

adisuissa replied:
Interesting. I don't have enough context on this test.
@htuch seems the test was added [here](https://github.com/envoyproxy/envoy/commit/73c46c9aba2c032f8fccb32e1b2c0c7a1df25a8d#diff-8e5f56463d7c4e0a70b111fef53b4e4d535b3c690b901441834fc4c891989167R138-R166)  (that you've reviewed and maybe you have some context), and we are wondering whether an xDS response of a single resource in delta-xDS can have both add/update and remove included?

htuch replied:
I don't think this was really deliberate AFAIU. It's been a long time so I don't have much context to add, but it seems a case where you could go either way with behaviors (ignore an unrelated or nonsensical removal, or perhaps a strong error like a NACK), this seems reasonable.

keithmattix replied:
Yeah testing that we don't NACK makes sense. I'm more inclined to keep this behavior for stability's sake and since the spec tends to be permissive of nonsensical removals

htuch replied:
I think this behavior is fine to retain, but please add a comment explaining what's going on for posterity.

adisuissa reviewed (commented):
Thanks for tackling this issue!
Overall I think this is ok, but I think more tests are needed.
Specifically I think we need to ensure that the owning cluster/listener are actually removing the resource prior to receiving a response with a `removed_resources`,
and tests that exercise the different validation paths.

keithmattix commented:
Thanks for the review @adisuissa!

> Specifically I think we need to ensure that the owning cluster/listener are actually removing the resource prior to receiving a response with a `removed_resources`,
> and tests that exercise the different validation paths.

I'm still getting up to speed on Envoy's testing approaches; is there a section of the integration tests that allow testing the dependent resource flow (specifically with SDS)?

keithmattix commented:
>  Specifically I think we need to ensure that the owning cluster/listener are actually removing the resource prior to receiving a response with a removed_resources,

I added an integration test that I think covers this; I'd appreciate feedback! I also modified the implementation so that removing a secret won't complete cluster warming.

adisuissa commented on test/integration/ads_integration_test.cc:
```diff
@@ -236,6 +236,93 @@ TEST_P(AdsIntegrationTest, ClusterInitializationUpdateOneOfThe2Warming) {
   test_server_->waitForGaugeGe("cluster_manager.active_clusters", 4);
 }
 
+TEST_P(AdsIntegrationTest, DeltaSdsRemovals) {
```
This test LGTM, but I wonder whether it should be here (ads_integration_test.cc) or in the sds integration tests.
Looking at the test it seems it only needs the sds to be updated, and if that's the case, I think it should be moved there.

Also, please add a short comment describing what the test validates and why it only relates to delta-xDS.

adisuissa replied:
Ah, I see.
LGTM, thanks!

adisuissa reviewed (commented):
Overall LGTM, thanks!
Left on comment on the organization of the integration test, but otherwise LGTM.

adisuissa commented:
Assigning Harvey as a senior-maintainer reviewer in the domain.
/assign @htuch

keithmattix commented on source/common/secret/sds_api.h:
```diff
@@ -92,7 +92,7 @@ class SdsApi : public Envoy::Config::SubscriptionBase<
 
 private:
-  absl::Status validateUpdateSize(int num_resources);
+  absl::Status validateUpdateSize(uint32_t added_resources_num, uint32_t removed_resources_num);
```
Done!

adisuissa commented on test/integration/ads_integration_test.cc:
```diff
@@ -236,6 +236,93 @@ TEST_P(AdsIntegrationTest, ClusterInitializationUpdateOneOfThe2Warming) {
 }
 
+TEST_P(AdsIntegrationTest, DeltaSdsRemovals) {
+  // This test is for delta only
```
Please add a comment why this is only relevant to Delta-xDS and not SotW.

Also comments style-guide, comments need to end with '.'.

keithmattix replied:
Ack; done!

adisuissa reviewed (dismissed):
LGTM, thanks!
One minor comment.

keithmattix commented:
/retest

keithmattix commented:
/retest

keithmattix commented:
/retest

keithmattix commented:
/retest

htuch commented on source/common/secret/sds_api.cc:
```diff
@@ -143,13 +143,24 @@ absl::Status SdsApi::onConfigUpdate(const std::vector<Config::DecodedResourceRef
+
+  if (removed_resources.size() == 1) {
+    // We only removed a resource here; don't go through the SotW init process
+    // and expect this to be cleaned up by the owning cluster or listener.
```
Please expand on the comment here. I think the key idea is this is a singleton resource subscription, so in normal operation, an xDS server should not be removing the resource until the referencing cluster / listener is no longer using it. 

@adisuissa why don't we structurally forbid singleton resource subscriptions from having removals in the baseline xDS infra, wondering why something like SDS needs to worry about this?

keithmattix replied:
+1 to enforcing this in the generic xDS machinery. I also think more documentation around "singleton" resources would be helpful as well; in any case, I've expanded the comment

htuch reviewed (commented):
I think this is right - if you look at what eds.cc is doing, it just ignores the removed resources for delta. LGTM modulo comments.

htuch reviewed (dismissed):
LGTM, thanks!
